src/File.hack

```diff
@@ -26,34 +26,63 @@ final class File {
     return Str\split($this->contents, "\n");
   }
 
+  /** 1-based line number */
   const type TLineNumber = int;
 
   /**
-   * Returns the 1-based line number of lint markers in the file.
+   * The lint error code to suppress, which could be either a linter name or a
+   * lint rule ID.
+   */
+  const type TErrorCode = string;
+
+  /**
+   * Returns the error codes to suppress according to lint markers in the file.
    */
   <<__Memoize>>
-  public function lintMarkersForLineBasedSuppression(
-  ): dict<string, keyset<self::TLineNumber>> {
-    return $this->getLines()
-      |> Vec\map_with_key(
+  public function errorCodesToSuppress(): shape(
+    'error_codes_in_file_lint_markers' => keyset<this::TErrorCode>,
+    'error_codes_in_line_lint_markers' => dict<this::TLineNumber, keyset<this::TErrorCode>>,
+  ) {
+    $all_matches = $this->getLines()
+      |> Vec\map(
         $$,
-        ($line_index, $line_content) ==>
-          Regex\every_match($line_content, re"/(?<marker_name>\w+)\[\w+\]/")
+        ($line_content) ==> Regex\every_match(
+          $line_content,
+          re"/(?<marker_name>\w+)\[(?<error_code>\w+)\]/",
+        )
           |> Vec\filter(
             $$,
             $match ==> LintMarkerName::isValid($match['marker_name']),
-          )
-          |> Vec\map($$, $match ==> tuple($line_index + 1, $match[0])),
-      )
-      |> Vec\flatten($$)
-      |> Dict\group_by($$, $line_marker ==> $line_marker[/* marker */ 1])
-      |> Dict\map(
-        $$,
-        $line_markers ==> Keyset\map(
-          $line_markers,
-          $line_marker ==> $line_marker[/* line number */ 0],
-        ),
+          ),
       );
+    $error_codes_in_file_lint_markers = () ==> {
+      foreach ($all_matches as $line_matches) {
+        foreach ($line_matches as $match) {
+          if ($match['marker_name'] === LintMarkerName::HHAST_IGNORE_ALL) {
+            yield $match['error_code'];
+          }
+        }
+      }
+    };
+    $error_codes_in_line_lint_markers = () ==> {
+      foreach ($all_matches as $line_index => $line_matches) {
+        if (!C\is_empty($line_matches)) {
+          $line_number = $line_index + 1;
+          $error_codes = () ==> {
+            foreach ($line_matches as $match) {
+              if ($match['marker_name'] !== LintMarkerName::HHAST_IGNORE_ALL) {
+                yield $match['error_code'];
+              }
+            }
+          };
+          yield $line_number => keyset($error_codes());
```

Comment:
remove generators, change back to map/filter pipelines.

1) it's much more idiomatic in hack
2) generator performance has not been a focus of optimization for years, and it shows - they are extremely ineffecient in HHVM compared to vec/dicts and the functional approach.

Reply:
It would be verbose without the functions in the `fb` namespace.

Reply:
Verbose implementation details is better than using generators. For example, https://github.com/hhvm/hhast/commit/50c3722b8a775a772302caacceae36b22505f3ad was a 2x speedup in 2017, and arrays have continued to be optimized - it's likely the problem is now worse.